lxd/apparmor.go

```diff
@@ -22,6 +23,27 @@ var aaEnabled = false
 
 var aaPath = shared.VarPath("security", "apparmor")
 
+const NESTING_AA_PROFILE = `
+  pivot_root,
+  mount /var/lib/lxd/shmounts/ -> /var/lib/lxd/shmounts/,
+  mount none -> /var/lib/lxd/shmounts/,
+  mount fstype=proc -> /usr/lib/x86_64-linux-gnu/lxc/**,
+  mount fstype=sysfs -> /usr/lib/x86_64-linux-gnu/lxc/**,
+  mount options=(rw,bind),
+  mount options=(rw,rbind),
+  deny /dev/.lxd/proc/** rw,
+  deny /dev/.lxd/sys/** rw,
```

Comment:
On Thu, Sep 10, 2015 at 03:17:52PM -0700, Serge Hallyn wrote:
> > @@ -22,6 +23,27 @@ var aaEnabled = false
> >  
> >  var aaPath = shared.VarPath("security", "apparmor")
> >  
> > +const NESTING_AA_PROFILE = `
> > +  pivot_root,
> > +  mount /var/lib/lxd/shmounts/ -> /var/lib/lxd/shmounts/,
> > +  mount none -> /var/lib/lxd/shmounts/,
> > +  mount fstype=proc -> /usr/lib/x86_64-linux-gnu/lxc/**,
> > +  mount fstype=sysfs -> /usr/lib/x86_64-linux-gnu/lxc/**,
> > +  mount options=(rw,bind),
> > +  mount options=(rw,rbind),
> > +  deny /dev/.lxd/proc/** rw,
> > +  deny /dev/.lxd/sys/** rw,
> 
> These mounts exist to bypass the kernel's enforcement that you cannot mount a new proc or sys in non-init userns if your only /proc or /sys is covered (which ours are - by /proc/cmdline etc, and /sys/devices/virtual/net).  So we mount uncovered versions under /dev/.lxd.

Ah ha, excellent, thanks.

> But we have apparmor rules to prevent use of some proc and sys files by pathname.  If proc and sys are mounted elsewhere, these rules are bypassed.
> 
> Really there's not much reason to have these:  first, so long as the container is unprivileged it can't read/write to root proc files anyway, and second the ability to mount them into containers makes the rules bypassable.  But it makes you work a little harder...

But this will go away once you can get rid of the "allow all mounting"
rule, I guess, so it's probably good to have. Thanks for the insight.

Tycho

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lxc/lxd/pull/1093/files#r39221576